core/src/main/java/hudson/Util.java

```diff
@@ -1662,7 +1666,27 @@ public static int permissionsToMode(Set<PosixFilePermission> permissions) {
             throw new IOException(e);
         }
     }
-
+    
+    /**
+     * Compute the number of calendar days elapsed since the given date.
+     * As it's only the calendar days difference that matter, "11.00pm" to "2.00am the day after" returns 1,
+     * even if there are only 3 hours between. As well as "10am" to "2pm" both on the same day, returns 0.
+     */
+    @Restricted(NoExternalUse.class)
+    public static int differenceInCalendarDay(@Nonnull Date a, @Nonnull Date b){
+        LocalDate aLocal = a.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
```

Comment:
I have half a preference for considering the user's time zone instead of the system's. (This is achievable by passing both dates down through to the jelly and letting it do this math instead.)

Reply:
IIRC HTTP Headers do not contain any information about timezone. You mean using JavaScript instead of server-side rendering ?

Reply:
Except for time zones that have managed to skip a day recently (which hasn't happened for decades or possibly centuries IIRC), this logic would work for any time zone.

Reply:
Yes, I meant JavaScript instead of server-side.

The issue isn't that the logic would work for any time zone, it's that the creature who wants to do the counting has a different perspective.

Date A is created at 1:03am New Zealand time.
https://www.timeanddate.com/worldclock/converter.html?iso=20171230T120300&p1=22&p2=195
Date B is created at 11:55pm New Zealand time on the same calendar date in New Zealand.
https://www.timeanddate.com/worldclock/converter.html?iso=20171231T105500&p1=22&p2=195

User C in New Zealand sees these two events as 0 days apart.

User D in Paris looks at these dates and sees them as 1 day apart.

Reply:
Oh, I see. Yeah, that would be an actual issue.

Reply:
To me it seems like we's want a more general fix for this, maybe [JENKINS-19887](https://issues.jenkins-ci.org/browse/JENKINS-19887) or the linked tickets there rather than introducing complexity here as a one-off case.

Reply:
Yeah, I'm fine with it not being addressed here. I just want people to be aware. Hence "my preference" as opposed to any stronger request.

The reason to mention it here is that it impacts the "api" perspective. If the values are both exposed to the client, then the client side can be upgraded to get this w/o needing to rebuild the "api".

OTOH, if jenkins treats the strings as the "api" and doesn't see them as "frozen", then evolving it later and doing it server-side is fine. -- I haven't built up a complete sense of how jenkins manages its "apis". (I just know that it has things that are apis.)

Reply:
@jsoref FYI about API:

As the code that is modified here is `@Restricted` (actually an inner class of a restricted class) we can consider it as an internal API (i.e. private) and so we could break it without any other consideration.

If there are public methods in a non-`@Restricted`, they could be used by other plugins / components, that's the developer API we need to care about in terms of binary compatibility.

Reply:
@Wadeck: k. that covers the java side, but what are the rules for the http side? (Maybe the answer is there aren't any constraints, which is probably a good thing here)

Reply:
@jsoref yeah in general I would say there is no constraint except that you need to keep consistency, security and compatibility in mind.